server/src/main/java/org/opensearch/indices/SystemIndices.java

```diff
@@ -147,6 +146,8 @@ private static CharacterRunAutomaton buildCharacterRunAutomaton(Collection<Syste
         Optional<Automaton> automaton = descriptors.stream()
             .map(descriptor -> Regex.simpleMatchToAutomaton(descriptor.getIndexPattern()))
             .reduce(Operations::union);
-        return new CharacterRunAutomaton(MinimizationOperations.minimize(automaton.orElse(Automata.makeEmpty()), Integer.MAX_VALUE));
+        return new CharacterRunAutomaton(
+            Operations.determinize(automaton.orElse(Automata.makeEmpty()), Operations.DEFAULT_DETERMINIZE_WORK_LIMIT)
```

Comment:
For some of these, it looks like we're 
1. parsing wildcard-type expressions with this simpleMatchToAutomaton, 
2. determinizing them (Operations.determinize) presumably just for the next step. 
3. tableizing them (CharRunAutomaton) for fastest execution.

This is a lot of investment, that makes sense if you are going to match gazillions of terms in the term dictionary.

But if you are just matching a couple things, maybe consider less heavy options?
1. `determinize()`, but don't tableize, and then match strings with `Operations.run()`
2. tableize, but don't `determinize()` by using `NFARunAutomaton`

Reply:
Yeah, for these cases it probably makes the most sense to run them as NFAs. Almost 100% of the time, they're just going to be prefix matches (or exact matches) anyway.

Reply:
The one area i'd say to take caution would be: multiple threads. So this is just a note, something to think about, as it requires some thought.

NFARunAutomaton is mutable and determinizes pieces "as-it-goes" and caches its work. This is different from existing RunAutomaton which is simple immutable arrays.

Reply:
Oh -- dang. Thanks a lot for the word of caution!

This `SystemIndices` is a singleton initialized on startup and gets called from all over the place. For example, in a `_bulk` request, each index name gets checked against the registered system index name patterns. Since we regularly process a bunch of bulk requests at once, the NFARunAutomaton would be a no-go.

It sounds like the `Operations.run()` approach is the one that won't blow up.

Of course, given that it's a one-time cost to determinize and tableize on startup, maybe that's the better choice if it shaves even a bit of effort at runtime?

Reply:
Well nothing there will blow up. but tableizing the thing does have a memory cost. Maybe just keep it the same, for now.

I need to followup and investigate how much performance these RunAutomatons even buy us anymore vs Operations.run(). The (non-tableized) automata representation has become much more efficient, thanks to @mikemccand.

Originally "Automaton" was a typical inefficient java representation, and "RunAutomaton" would tableize it into efficient array, which gives you e.g. cache locality and all that. But now Automaton has that locality, it is represented by two arrays: `states[]` and `transitions[]`. 

So it is possible the entire RunAutomaton is just a waste of code and RAM and that we could nuke it: needs investigation.